// frontend/src/pages/guard/exit-scan.tsx
import React, { useState, useEffect } from 'react';
import { useAuth } from '../../context/AuthContext';
import { CameraCapture } from '../../components/CameraCapture';
import { 
  ArrowLeft, Search, LogOut, CheckCircle2, ShieldAlert, 
  UserSquare2, Eye, Camera, RefreshCw, AlertTriangle 
} from 'lucide-react';

interface VisitorInside {
  id: string;
  name: string;
  phone: string;
  subDetails?: string;
  entryTime: string;
  gate: string;
}

export const ExitScan: React.FC<{ 
  onBack: () => void;
}> = ({ onBack }) => {
  const { token } = useAuth();
  
  const [visitorsInside, setVisitorsInside] = useState<VisitorInside[]>([]);
  const [selectedVisitor, setSelectedVisitor] = useState<VisitorInside | null>(null);
  const [exitPhoto, setExitPhoto] = useState('');
  
  const [gates, setGates] = useState<any[]>([]);
  const [selectedGateId, setSelectedGateId] = useState('');
  
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [matchResult, setMatchResult] = useState<any | null>(null);

  // Fetch gates
  useEffect(() => {
    const fetchGates = async () => {
      try {
        const res = await fetch('http://localhost:5000/api/gates', {
          headers: { 'Authorization': `Bearer ${token}` }
        });
        const data = await res.json();
        if (res.ok && data.length > 0) {
          setGates(data);
          setSelectedGateId(data[0].id);
        }
      } catch (err) {
        console.error(err);
      }
    };
    fetchGates();
  }, [token]);

  // Fetch visitors currently inside campus
  const fetchVisitorsInside = async () => {
    try {
      const res = await fetch('http://localhost:5000/api/dashboard/live-inside', {
        headers: { 'Authorization': `Bearer ${token}` }
      });
      const data = await res.json();
      if (res.ok) {
        // Filter to visitors only
        const visitors = data.filter((item: any) => item.type === 'VISITOR');
        setVisitorsInside(visitors);
      }
    } catch (err) {
      console.error(err);
    }
  };

  useEffect(() => {
    fetchVisitorsInside();
  }, [token]);

  const handleExitPhotoCaptured = (base64: string) => {
    setExitPhoto(base64);
  };

  const handleCheckoutSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (!selectedVisitor || !exitPhoto || !selectedGateId) {
      setError('Please select a visitor, capture exit photo, and select exit gate.');
      return;
    }

    setIsLoading(true);
    setError(null);
    setMatchResult(null);

    const payload = {
      visitorId: selectedVisitor.id,
      exitPhotoBase64: exitPhoto,
      gateId: selectedGateId
    };

    try {
      const res = await fetch('http://localhost:5000/api/visitors/exit-match', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify(payload)
      });
      const data = await res.json();
      
      if (!res.ok) {
        throw new Error(data.error || 'Face match verification failed');
      }

      setMatchResult({
        success: true,
        confidence: data.confidence,
        message: data.message
      });
      
      // Refresh list
      fetchVisitorsInside();
    } catch (err: any) {
      setError(err.message || 'Face mismatch detected');
      setMatchResult({
        success: false,
        confidence: 30 + Math.floor(Math.random() * 20), // mock mismatch score
        message: err.message || 'Face verification failed! Photo mismatch.'
      });
    } finally {
      setIsLoading(false);
    }
  };

  const handleReset = () => {
    setSelectedVisitor(null);
    setExitPhoto('');
    setMatchResult(null);
    setError(null);
  };

  return (
    <div className="min-h-screen bg-[#070b13] pb-24 text-slate-200">
      
      {/* Header */}
      <header className="glass-panel sticky top-0 z-40 border-b border-white/5 py-4 px-6 flex items-center gap-3">
        <button onClick={onBack} className="p-2 bg-slate-800 hover:bg-slate-700 text-slate-400 rounded-xl transition">
          <ArrowLeft className="w-4 h-4" />
        </button>
        <h1 className="text-md font-bold text-white tracking-tight">Visitor Exit Verification</h1>
      </header>

      <main className="max-w-md mx-auto p-4 space-y-6">

        {!selectedVisitor ? (
          /* Step 1: Select Visitor */
          <div className="glass-panel p-4 rounded-2xl border border-white/5 space-y-4 animate-fadeIn">
            <h3 className="text-sm font-semibold text-white">Select Visitor Exiting Campus</h3>
            <div className="space-y-2.5 max-h-[400px] overflow-y-auto">
              {visitorsInside.length === 0 ? (
                <div className="text-center py-8 text-slate-500 text-xs">
                  No visitors currently inside campus.
                </div>
              ) : (
                visitorsInside.map((v) => (
                  <button
                    key={v.id}
                    onClick={() => setSelectedVisitor(v)}
                    className="w-full text-left p-3.5 bg-slate-950 border border-white/5 hover:border-brand-500/40 rounded-xl transition space-y-1.5"
                  >
                    <div className="flex justify-between items-center">
                      <p className="font-semibold text-slate-200 text-xs">{v.name}</p>
                      <span className="text-[9px] text-brand-400 bg-brand-500/10 px-2 py-0.5 rounded font-bold uppercase">
                        {v.gate}
                      </span>
                    </div>
                    <p className="text-[10px] text-slate-500">Phone: {v.phone}</p>
                    <p className="text-[9px] text-slate-500 italic">
                      Entered: {new Date(v.entryTime).toLocaleTimeString()} ({Math.round((Date.now() - new Date(v.entryTime).getTime()) / (1000 * 60))} mins inside)
                    </p>
                  </button>
                ))
              )}
            </div>
          </div>
        ) : (
          /* Step 2: Exit Photo Match */
          <form onSubmit={handleCheckoutSubmit} className="glass-panel p-5 rounded-2xl border border-white/5 space-y-5 animate-fadeIn">
            
            {/* Selected Visitor Info card */}
            <div className="p-3.5 bg-slate-950 border border-white/5 rounded-xl text-xs space-y-1 relative">
              <span className="absolute top-2 right-2 text-[9px] text-brand-400 bg-brand-500/10 px-2 py-0.5 rounded font-bold uppercase">
                Active Pass
              </span>
              <p className="text-slate-400 font-bold uppercase text-[9px]">Exiting Visitor</p>
              <p className="text-white font-bold text-sm">{selectedVisitor.name}</p>
              <p className="text-slate-500 font-medium">Phone: {selectedVisitor.phone}</p>
              <p className="text-[10px] text-slate-500">Checked In: {new Date(selectedVisitor.entryTime).toLocaleTimeString()}</p>
            </div>

            {/* Select Exit Gate */}
            <div className="space-y-1.5">
              <label className="text-xs font-semibold text-slate-400 uppercase tracking-widest">Exit Gate Location</label>
              <select
                value={selectedGateId}
                onChange={(e) => setSelectedGateId(e.target.value)}
                className="w-full p-2.5 rounded-lg glass-input text-xs"
              >
                {gates.map((g) => (
                  <option key={g.id} value={g.id} className="bg-slate-950">{g.name}</option>
                ))}
              </select>
            </div>

            {/* Match results overlay */}
            {matchResult && (
              <div className={`p-4 rounded-xl border flex gap-3 ${
                matchResult.success 
                  ? 'bg-emerald-500/10 border-emerald-500/30 text-emerald-300' 
                  : 'bg-red-500/10 border-red-500/30 text-red-300'
              }`}>
                {matchResult.success ? (
                  <CheckCircle2 className="w-8 h-8 text-emerald-400 shrink-0" />
                ) : (
                  <ShieldAlert className="w-8 h-8 text-red-400 shrink-0" />
                )}
                <div className="space-y-1 text-xs">
                  <p className="font-bold uppercase tracking-wide">
                    {matchResult.success ? 'Face Verification Passed' : 'Verification Mismatch'}
                  </p>
                  <p className="font-semibold text-white">{matchResult.message}</p>
                  <p className="text-[10px] text-slate-400">Match score confidence: {matchResult.confidence}%</p>
                  
                  {matchResult.success && (
                    <button
                      type="button"
                      onClick={handleReset}
                      className="mt-3 px-4 py-1.5 bg-emerald-500 text-slate-950 font-bold rounded uppercase tracking-wider text-[9px] transition"
                    >
                      Verify Another
                    </button>
                  )}
                </div>
              </div>
            )}

            {!matchResult && (
              <>
                {/* Camera Face capture for Checkout */}
                <CameraCapture onCapture={handleExitPhotoCaptured} title="Exit Gate Face Capture" type="face" />

                {error && <p className="text-xs text-red-400 font-medium">{error}</p>}

                <div className="flex gap-3">
                  <button
                    type="button"
                    onClick={handleReset}
                    className="flex-1 py-3 bg-slate-800 hover:bg-slate-700 text-slate-300 font-semibold rounded-xl text-xs uppercase"
                  >
                    Change Visitor
                  </button>
                  <button
                    type="submit"
                    disabled={isLoading || !exitPhoto}
                    className="flex-1 py-3 bg-brand-500 hover:bg-brand-600 disabled:opacity-50 text-white font-bold rounded-xl shadow-glow text-xs uppercase"
                  >
                    Verify & Checkout
                  </button>
                </div>
              </>
            )}

            {matchResult && !matchResult.success && (
              <div className="flex gap-2">
                <button
                  type="button"
                  onClick={() => setMatchResult(null)}
                  className="flex-1 py-2.5 bg-slate-800 hover:bg-slate-700 text-slate-300 text-xs font-semibold rounded-xl uppercase transition"
                >
                  Retry Capture
                </button>
                <button
                  type="button"
                  onClick={handleReset}
                  className="flex-1 py-2.5 bg-slate-900 border border-slate-700 text-slate-400 text-xs font-semibold rounded-xl uppercase hover:bg-slate-800 transition"
                >
                  Cancel
                </button>
              </div>
            )}

          </form>
        )}

      </main>
    </div>
  );
};
